### DISCUSSION TITLE
Yubikey

### DISCUSSION THREAD

**Discussion by onyxcherry** [dorssel/usbipd-win#127] [Q&A] [answered]
Hello,
how do you use the Yubikey basic functionality in e.g. Ubuntu-20.04, especially with what (USB) drivers to get it working?

Currently (`sudo lsusb -v` inside WSL of course):
```
Bus 001 Device 002: ID 1050:0120 Yubico.com Yubikey Touch U2F Security Key
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0
  bDeviceSubClass         0
  bDeviceProtocol         0
  bMaxPacketSize0        64
  idVendor           0x1050 Yubico.com
  idProduct          0x0120 Yubikey Touch U2F Security Key
  bcdDevice            5.26
  iManufacturer           1 Yubico
  iProduct                2 Security Key by Yubico
  iSerial                 0
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0029
    bNumInterfaces          1
    bConfigurationValue     1
    iConfiguration          0
    bmAttributes         0x80
      (Bus Powered)
    MaxPower               30mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           2
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0
      bInterfaceProtocol      0
      iInterface              0
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.10
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      34
         Report Descriptors:
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x04  EP 4 OUT
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               2
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x84  EP 4 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               2
can't get device qualifier: Resource temporarily unavailable
can't get debug descriptor: Resource temporarily unavailable
Device Status:     0x0000
  (Bus Powered)
```
and no output from `CtapHidDevice.list_devices()` (python-fido2 as root).

---

`cat /etc/udev/rules.d/90-fido2-own.rules`:
```
#Udev rule for allowing HID access to Yubico devices for FIDO support.

KERNEL=="hidraw*", SUBSYSTEM=="hidraw", \
  MODE="0664", GROUP="plugdev", ATTRS{idVendor}=="1050"
```

Reloading udev with `sudo udevadm trigger` or even restarting the Windows (host) computer doesn't result in working :(

**Comment by dorssel** [comment, marked as answer]
For the HID interface, see #90.
For the other interface (smartcard, etc.) you will need to compile a kernel with the correct drivers, I think.

**Reply by onyxcherry** [reply]
So no one of FIDO2 features (e.g. generating ssh key with RK option) or simply - basic - WebAuthn or U2F could work?

What are the modules needed to compile into a new kernel?
It is a lack of drivers, isn't it?

That's obviously an open question, future finding are welcome 😄

**Reply by dorssel** [reply]
I just did a `gpg --card-status` for my Yubikey 5 on a regular Linux Hyper-V VM.
It loaded `usbhid` and `input_leds`.

**Comment by mcdonaldseanp** [comment]
I ran in to this problem or at least to me it looked like a similar set of symptoms: can connect yubikey from the windows side to WSL with usbipd, see it with `lsusb` but wasn't able to use it with gpg. I'm not sure about any other functionality like WebAuthn but I was at least able to get yubikey working together with gpg (including using gpg-agent instead of ssh-agent to access hosts with keys unlocked with my Yubikey).

For me it turned out the issue was that the `pcscd` [service](https://linux.die.net/man/8/pcscd) wasn't running. I was able to get things functional with my yubikey including running things with gpg by starting that service: `sudo service pcscd start`. If I remove the key from my workstation and need to use it again I have to restart pcscd every time (this is probably just my own lack of knowledge with pcscd though), but it takes so little time to restart it doesn't bother me.

**Reply by onyxcherry** [reply]
Unfortunately, I have an Blue one - it hasn't GPG support.

I attached the key by `sudo /usr/lib/linux-tools/5.4.0-77-generic/usbip attach -r 172.23.48.1 -b 1-2` and saw it in `lsusb` output.
But in `dmesg` output I see:
```
vhci_hcd vhci_hcd.0: pdev(0) rhport(0) sockfd(3)
vhci_hcd vhci_hcd.0: devid(65538) speed(2) speed_str(full-speed)
vhci_hcd vhci_hcd.0: Device attached
vhci_hcd: vhci_device speed not set
usb 1-1: new full-speed USB device number 2 using vhci_hcd
vhci_hcd: vhci_device speed not set
usb 1-1: SetAddress Request (2) to port 0
usb 1-1: New USB device found, idVendor=1050, idProduct=0120, bcdDevice= 5.26
usb 1-1: New USB device strings: Mfr=1, Product=2, SerialNumber=0
usb 1-1: Product: Security Key by Yubico
usb 1-1: Manufacturer: Yubico
hid-generic 0003:1050:0120.0001: device has no listeners, quitting
```

How to set the speed?

Edit:
`/sys/bus/usb/devices/1-1/speed` consists `12`
`.../product` consists `Security Key by Yubico` so it's a right directory

Edit °2:
`gzip -d < /proc/config.gz | grep HIDDEV` outputs `# CONFIG_USB_HIDDEV is not set` so maybe building the kernel with `CONFIG_USB_HIDDEV` is the solution?

**Reply by dorssel** [reply]
The `vhci_hcd: vhci_device speed not set` line is perfectly normal. There is no need to set the speed. This is reported by all USBIP devices on Linux, also for Linux-to-Linux so not `usbipd-win` related.

And yes, for YubiKey you will need HID drivers. I myself do not use WSL, but I have my YubiKey working under Hyper-V. See my earlier post. It uses the `usbhid` driver.

**Reply by onyxcherry** [reply]
[device speed] Ok, fine 😄 

Does anyone made this work?

dmesg shows the following when the browser wants to use an Yubikey:
```
hid-generic 0003:1050:0120.0001: device has no listeners, quitting
dxgk:err: is_handle_valid Entry is freed 40000080 2
dxgk:err: hmgrtable_get_object_by_type invalid handle 40000080
              <--- snip --->
dxgk:err: hmgrtable_get_object_by_type invalid handle 40000080
dxgk:err: adapter_by_handle failed 40000080
TCP: eth0: Driver has suspect GRO implementation, TCP performance may be compromised.
dxgk:err: is_handle_valid Entry is freed 40000080 2
dxgk:err: hmgrtable_get_object_by_type invalid handle 40000080
dxgk:err: adapter_by_handle failed 40000080
```

In other hand, I have `libudev-dev`, `libfido2-1`, `libfido2-dev` installed.

And probably have required drivers:
```
tomasz@wsl:/sys/bus/usb/drivers/usbhid$ ls -la
1-1:1.0 -> ../../../../devices/platform/vhci_hcd.0/usb1/1-1/1-1:1.0/
bind
module -> ../../../../module/usbhid/
new_id
remove_id
uevent
unbind
```
Nothing shows at `/dev/console` 😢

**Reply by dorssel** [reply]
If you do `sudo cat /dev/console` and you touch your YubiKey, does it give any response?

**Reply by onyxcherry** [reply]
Nothing 🙄

**Reply by dorssel** [reply]
EDIT: strike that previous post ... that was not the YubiKey...

**Reply by dorssel** [reply]
What does the `usbhid` driver node say? Mine says:

```
/sys/bus/usb/drivers/usbhid$ ls -la
total 0
drwxr-xr-x 2 root root    0 Dec 20 16:52 .
drwxr-xr-x 6 root root    0 Dec 20 09:44 ..
lrwxrwxrwx 1 root root    0 Dec 20 17:37 1-1:1.0 -> ../../../../devices/platform/vhci_hcd.0/usb1/1-1/1-1:1.0
lrwxrwxrwx 1 root root    0 Dec 20 17:37 1-1:1.1 -> ../../../../devices/platform/vhci_hcd.0/usb1/1-1/1-1:1.1
--w------- 1 root root 4096 Dec 20 17:27 bind
lrwxrwxrwx 1 root root    0 Dec 20 17:27 module -> ../../../../module/usbhid
-rw-r--r-- 1 root root 4096 Dec 20 17:27 new_id
-rw-r--r-- 1 root root 4096 Dec 20 17:27 remove_id
--w------- 1 root root 4096 Dec 20 16:52 uevent
--w------- 1 root root 4096 Dec 20 17:27 unbind
/sys/bus/usb/drivers/usbhid$
```

**Reply by onyxcherry** [reply]
```
tomasz@wsl:/sys/bus/usb/drivers/usbhid$ ls -la
total 0
drwxr-xr-x 2 root root    0 Dec 20 17:40 .
drwxr-xr-x 8 root root    0 Dec 20 17:40 ..
lrwxrwxrwx 1 root root    0 Dec 20 17:40 1-1:1.0 -> ../../../../devices/platform/vhci_hcd.0/usb1/1-1/1-1:1.0
--w------- 1 root root 4096 Dec 20 17:40 bind
lrwxrwxrwx 1 root root    0 Dec 20 17:40 module -> ../../../../module/usbhid
-rw-r--r-- 1 root root 4096 Dec 20 17:40 new_id
-rw-r--r-- 1 root root 4096 Dec 20 17:40 remove_id
--w------- 1 root root 4096 Dec 20 17:40 uevent
--w------- 1 root root 4096 Dec 20 17:40 unbind
```

**Reply by nwayve** [reply]
I ran into this issue recently.  After attaching the yubikey usb to wsl via `uspipd wsl attach --busid {id}`, the device would show up in wsl in `lsusb` but not in `ykman list`.  The command `sudo service pcscd start` fixed it initially, but after restarting Windows and getting back into the wsl shell, the issue happened again but the service start command didn't work this time.  I had to restart the service with `sudo service pcscd restart` in order for it to work again.

**Comment by damex** [comment]
I hit the same problem with Debian 11 trying to use same Yubikey for OTP. usbip passthrough successfully and PIV works fine. attempting to check otp by 'ykman otp info' results in saying that otp interface is not found and touching Yubikey generate no otp. not sure if there is a workaround for that.

**Reply by dorssel** [reply]
On a regular, current WSL kernel I get:

```
$ sudo ykman otp info
Slot 1: programmed
Slot 2: empty 
```

and

```
$ sudo cat /dev/console
<***touch***>
ccccccnjkbbrjuvh<***snip***>
```

This is without any further config or extra drivers.

**Reply by davidshen84** [reply]
Hi, I just started using this tool today and have managed to get my Yubikey works with SSH and GPG. Now want to configure my Yubikey to work with my WSL's `sudo` command. But I cannot configure the keys using `pamu2fcfg`. It says it cannot find any U2F devices. I wonder if it is also HID related.

```
zgrep -i hiddev /proc/config.gz
# CONFIG_USB_HIDDEV is not set
```

**Reply by dorssel** [reply]
That makes sense. When I install with `sudo apt install pamu2fcfg`, it sucks in `libhidapi-hidraw0`. That suggests that HIDRAW support is required. Seems you need to compile a kernel with that included.

**Reply by davidshen84** [reply]
Any reason why `CONFIG_USB_HIDDEV` is not enabled in the default kernel? It is an essential feature in Linux.

**Reply by dorssel** [reply]
Not for WSL, I guess.
If you think it should be enabled by default you can request it at https://github.com/microsoft/WSL/issues.

**Reply by davidshen84** [reply]
OK, let's poke them... https://github.com/microsoft/WSL/issues/8302

**Reply by Seldaek** [reply]
@davidshen84 do you have any instructions how you got it to work with SSH? 

I got as far as what @dorssel wrote above, `sudo cat /dev/console` works when I tap the key (YubiKey 5 Nano), `ykman info` sees the key, but `ykman otp info` gives `Error: No YubiKey found with the given interface(s)`.

**Reply by davidshen84** [reply]
`ykman` doesn't work because of WSL kernel config. To use the key in WSL for SSH connection, please check out this [guide][1]. You can ignore the gpg part in it.

[1]: https://github.com/drduh/YubiKey-Guide#ssh

**Reply by onyxcherry** [reply]
Or you can build the kernel [https://github.com/dorssel/usbipd-win/wiki/WSL-support#building-your-own-usbip-enabled-wsl-2-kernel] on your own and have FIDO2 working :D

**Reply by Seldaek** [reply]
@onyxcherry is it correct that you still have to mount the USB from windows and then cannot use it anymore from windows? Because if that's the case I guess I rather skip this whole endeavor.

**Reply by onyxcherry** [reply]
You are entirely right about that process

**Comment by ghost** [comment]
Can someone confirm if I should be able to generate an `ed25519-sk`  or `ecdsa-sk` key? I have tried but failed. I also compiled my own kernel with `usbhid` driver, but no joy.

**Reply by onyxcherry** [reply]
# It works

#### However, with and only with custom built

Windows (host):

```cmd
C:\WINDOWS\system32>usbipd list
Present:
BUSID  DEVICE                                                        STATE
[...]
1-2    Input USB device                                       Attached
[...]
```

---

WSL2 (with the custom built kernel):
```plain
root@TomekAspire5:~# lsusb -v 2>/dev/null | grep -A2 Yubico
Bus 001 Device 002: ID 1050:0120 Yubico.com Yubikey Touch U2F Security Key
Device Descriptor:
  bLength                18
--
  idVendor           0x1050 Yubico.com
  idProduct          0x0120 Yubikey Touch U2F Security Key
  bcdDevice            5.26
  iManufacturer           1 Yubico
  iProduct                2 Security Key by Yubico
  iSerial                 0
  bNumConfigurations      1
```

(here is a firmware version greater than 5.2.3 which is the minimum one to [support](https://support.yubico.com/hc/en-us/articles/360016649319-YubiKey-5-2-3-Enhancements-to-FIDO-2-Support) FIDO2)

and last but not least - the kernel built configuration:
```bash
root@TomekAspire5:~# zcat /proc/config.gz | grep USB | grep -v "is not set"
CONFIG_USB_NET_DRIVERS=y
CONFIG_USB_USBNET=y
CONFIG_USB_NET_CDCETHER=y
CONFIG_USB_NET_CDC_NCM=y
# USB HID support
CONFIG_USB_HID=y
CONFIG_USB_HIDDEV=y
# end of USB HID support
CONFIG_USB_OHCI_LITTLE_ENDIAN=y
CONFIG_USB_SUPPORT=y
CONFIG_USB_COMMON=y
CONFIG_USB_ARCH_HAS_HCD=y
CONFIG_USB=y
CONFIG_USB_ANNOUNCE_NEW_DEVICES=y
# Miscellaneous USB options
CONFIG_USB_DEFAULT_PERSIST=y
CONFIG_USB_AUTOSUSPEND_DELAY=2
# USB Host Controller Drivers
# USB Device Class drivers
CONFIG_USB_ACM=y
# NOTE: USB_STORAGE depends on SCSI but BLK_DEV_SD may
# also be needed; see USB_STORAGE Help for more info
# USB Imaging devices
CONFIG_USBIP_CORE=y
CONFIG_USBIP_VHCI_HCD=y
CONFIG_USBIP_VHCI_HC_PORTS=8
CONFIG_USBIP_VHCI_NR_HCS=1
CONFIG_USBIP_DEBUG=y
# USB port drivers
CONFIG_USB_SERIAL=y
CONFIG_USB_SERIAL_CH341=y
CONFIG_USB_SERIAL_CP210X=y
CONFIG_USB_SERIAL_FTDI_SIO=y
# USB Miscellaneous drivers
# USB Physical Layer drivers
# end of USB Physical Layer drivers
```

and voilà,
```plain
root@TomekAspire5:~# sudo ssh-keygen -t ed25519-sk
Generating public/private ed25519-sk key pair.
You may need to touch your authenticator to authorize key generation.
Enter PIN for authenticator:
You may need to touch your authenticator (again) to authorize key generation.
Enter file in which to save the key (/root/.ssh/id_ed25519_sk):
```

**Reply by ghost** [reply]
Thanks! I had to use `sudo` to `ssh-keygen -t ed25519-sk` because I am using Fedoraremix WSL and `udev` service does not work.

However, I managed to get `udev` running using [Genie](https://github.com/arkane-systems/genie) and using the following [udev rules](https://raw.githubusercontent.com/Yubico/libfido2/main/udev/70-u2f.rules) (just had to change the `GROUP="plugdev"` to the one I use)

**Reply by 1-bit-wonder** [reply]
I was able to get this working too, I wrote a [tutorial ](https://1-bit-wonder.github.io/blog/how-to-use-yubikey-with-wsl/How%20to%20use%20Yubikey%20with%20WSL%20via%20USB%20passthrough/) for anyone interested.

**Reply by returntrip** [reply]
It seems like the custom kernel is not required anymore since the last version of WSL2? Can anyone else confirm?

I removed my custom `.wslconfig` and I can still use the yubikey

**Reply by dorssel** [reply]
Some functions (e.g. gpg) now work out-of-the-box. The latest default WSL kernel has
```
#
# HID support
#
CONFIG_HID=y
# CONFIG_HID_BATTERY_STRENGTH is not set
# CONFIG_HIDRAW is not set
# CONFIG_UHID is not set
CONFIG_HID_GENERIC=y
```

and

```
#
# USB HID support
#
CONFIG_USB_HID=y
# CONFIG_HID_PID is not set
# CONFIG_USB_HIDDEV is not set
# end of USB HID support
```

I believe FIDO and some other functions also require CONFIG_HIDRAW and/or CONFIG_USB_HIDDEV. Not sure though...

**Reply by returntrip** [reply]
@dorssel confirmed gpg works OOB whilst FIDO needs extra kernel flags
